test/data/mock.php

```diff
@@ -0,0 +1,241 @@
+		if ( isset( $req->query['json'] ) ) {
+			header( 'Content-Type: application/json' );
+			echo '{ "code": 40, "message": "Bad Request" }';
+		} else {
+			echo 'plain text message';
+		}
+	}
+
+	protected function headers( $req ) {
+		header( 'Sample-Header: Hello World' );
+		header( 'Empty-Header: ' );
+		header( 'Sample-Header2: Hello World 2' );
+
+		foreach ( explode( '|' , $req->query[ 'keys' ] ) as $key ) {
+			// Only echo if key exists in the header
+			if ( isset( $req->headers[ strtoupper( $key ) ] ) ) {
+				echo "$key: " . $req->headers[ strtoupper( $key ) ] . "\n";
+			}
+		}
+
+	}
+
+	protected function echoData( $req ) {
+		echo file_get_contents('php://input');
+	}
+
+	protected function echoQuery( $req ) {
+		echo $_SERVER['QUERY_STRING'];
+	}
+
+	protected function echoMethod( $req ) {
+		echo $req->method;
+	}
+
+	protected function echoHtml( $req ) {
+		header( 'Content-type: text/html' );
+		echo '<div id="method">' . $req->method . '</div>';
+		echo '<div id="query">' . $_SERVER['QUERY_STRING'] . '</div>';
+		echo '<div id="data">' . file_get_contents('php://input') . '</div>';
+	}
+
+	protected function etag( $req ) {
+		$hash = md5( $req->query['ts'] );
+		$etag = 'W/"' . $hash . '"';
+
+		$ifNoneMatch = isset( $req->headers['IF-NONE-MATCH'] ) ? $req->headers['IF-NONE-MATCH'] : '';
+		if ($ifNoneMatch === $etag) {
+			header('HTTP/1.0 304 Not Modified');
+			return;
+		}
+
+		header("Etag: $etag");
+		echo "ETag: $etag\n";
+		if ( $ifNoneMatch ) {
+			echo "If-None-Match: $ifNoneMatch\n";
+		}
+	}
+
+	protected function ims( $req ) {
+		$ts = $req->query['ts'];
+
+		$ims = isset( $req->headers['IF-MODIFIED-SINCE'] ) ? $req->headers['IF-MODIFIED-SINCE'] : '';
+		if ($ims === $ts) {
+			header('HTTP/1.0 304 Not Modified');
+			return;
+		}
+
+		header("Last-Modified: $ts");
+		echo "Last-Modified: $ts\n";
+		if ( $ims ) {
+			echo "If-Modified-Since: $ims\n";
+		}
+	}
+
+	protected function status( $req ) {
+		header( "HTTP/1.0 {$req->query['code']} {$req->query['text']}" );
+	}
+
+	protected function testHTML( $req ) {
+		header( 'Content-type: text/html' );
+		$html = file_get_contents( __DIR__ . '/test.include.html' );
+		$html = str_replace( '{{baseURL}}', $req->query['baseURL'], $html );
+		echo $html;
+	}
+
+	protected function cspFrame( $req ) {
+		// This is CSP only for browsers with "Content-Security-Policy" header support
+		// i.e. no old WebKit or old Firefox
+		header( "Content-Security-Policy: default-src 'self'; report-uri ./mock.php?action=cspLog" );
+		header( 'Content-type: text/html' );
+		echo file_get_contents( __DIR__ . '/csp.include.html' );
+	}
+
+	protected function cspLog( $req ) {
+		file_put_contents( $this->cspFile, 'error' );
+	}
+
+	protected function cspClean( $req ) {
+		file_put_contents( $this->cspFile, '' );
+		unlink( $this->cspFile );
+	}
+
+	public function __construct() {
+		$this->cspFile = __DIR__ . '/support/csp.log';
+	}
+
+	public function respond( stdClass $req ) {
```

Comment:
I've already narrowed down the number of different response handlers from what it originally was. Any additional improvements can be done later instead of blocking this pull.

I do not, however, that we don't currently have cases where we need multiple of `wait`, `header` and `content`. Indicated by there not being any such method right now that does multiple. It's pretty well-isolated. Making it more generic and changing the actual fixtures used would be better, but that also requires more careful examination of what the tests are actually testing and what they need. For example, some tests probably don't need the exact output they get right now.